### ISSUE TITLE
Recursive type aliases don't work as expected

### ISSUE THREAD

**Issue by teresaejunior** [microsoft/pyright#1484]
Hi! First of all, thanks for Pyright! Although it is just supposed to be a type checker, it is the only linter that finds real errors (even beyond typing) most of the time, while all other I tried mostly waste my time with false alarms.

I was having problems trying to make my recursive types work, but then I noticed the example from [devblogs.microsoft.com](https://devblogs.microsoft.com/python/pylance-introduces-five-new-features-that-enable-type-magic-for-python-developers/) doesn't either.

```python
from typing import Dict, List, Union
ConfigValue = Union[
    bool, 
    str, 
    float,
    List["ConfigValue"],
    Dict[str, "ConfigValue"],
]
config: List[ConfigValue] = [
    {"python.experiments.optInto":        
        ["Experiment1", "Experiment13", "Experiment56", "Experiment106"]},
    {"python.formatting.provider": "black"},
    {"python.formatting.blackArgs": ["--line-length", "130"]},
    {"python.sortImports.args": ["-rc", "-sp isort.cfg"]},
]
config[1]["python.formatting.provider"]
```
```
$ pyright test.py 
No configuration file found.
stubPath /home/teresaejunior/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/home/teresaejunior/test.py
  21:1 - error: "__getitem__" method not defined on type "bool" (reportGeneralTypeIssues)
  21:1 - error: Argument of type "Literal['python.formatting.provider']" cannot be assigned to parameter "i" of type "int | slice" in function "__getitem__"
    Type "Literal['python.formatting.provider']" cannot be assigned to type "int | slice"
      "Literal['python.formatting.provider']" is incompatible with "int"
      "Literal['python.formatting.provider']" is incompatible with "slice" (reportGeneralTypeIssues)
  21:1 - error: No overloads for "__getitem__" match the provided arguments
    Argument types: (Literal['python.formatting.provider']) (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 infos 
Completed in 0.605sec
```
My personal code, slightly different:
```py
from typing import Dict, Union
Conf = Dict[str, Union[int, 'Conf']]
obj: Conf = {'a': 1, 'b': {'c': 1, 'd': {'e': 1}}}
obj['b']['c']
```
```
No configuration file found.
stubPath /home/teresaejunior/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/home/teresaejunior/test.py
  4:1 - error: "__getitem__" method not defined on type "int" (reportGeneralTypeIssues)
  4:1 - error: Object of type "Conf" is not subscriptable (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 infos 
Completed in 0.576sec
```
**VS Code extension or command-line**
pyright@1.1.110 or Pylance 2021.2.2.

**Comment by erictraut** [comment]
You have added a line of code to the example provided in the blog: `config[1]["python.formatting.provider"]`. The line that you've added is buggy, and Pyright is highlighting the bug in your code. If you change the code to the following, it will work correctly.

```python
config_entry = config[1]
if isinstance(config_entry, dict):
    print(config_entry["python.formatting.provider"])
```

Likewise, the second code sample you provided has a bug that is being highlighted by pyright. The type of the expression `obj['b']` is a union of `int` and `Conf`. If it's an `int`, you will receive a runtime exception if you attempt to index it because the `int` class does not define a `__getitem__` method.

You would need to add more checks at runtime to prevent this exception. Something like this...
```python
value = obj["b"]
if isinstance(value, int):
    print("Value is int")
else:
    print("Value is dict")
    print(value["c"])
```

**Comment by teresaejunior** [comment]
Well, it isn't the first time I thought I had found a bug in Pyright, and it was my fault. I guess I'm just not used yet to typing in Python. My issue here is I thought Union was supposed to be used implicitly. Thanks for your help!
